src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -1429,15 +1429,36 @@ public static class PipelineStageDefinitionBuilder
         }
 
         // private methods
-        private static bool IsTConnectToOrEnumerableTConnectTo<TConnectFrom, TConnectTo>()
+        private static bool AreGraphLookupTypesCompatible<TConnectFrom, TConnectTo>()
         {
             if (typeof(TConnectFrom) == typeof(TConnectTo))
             {
                 return true;
             }
 
+            var connectFromInterfaces = typeof(TConnectFrom).GetTypeInfo().GetInterfaces().Where(i => i.GetTypeInfo().IsGenericType);
+            var connectToInterfaces = typeof(TConnectTo).GetTypeInfo().GetInterfaces().Where(i => i.GetTypeInfo().IsGenericType);
+            if (connectFromInterfaces.Any(i => connectToInterfaces.Contains(i)))
+            {
+                return true;
+            }
+            if (connectFromInterfaces.Contains(typeof(TConnectTo)))
+            {
+                return true;
+            }
+            if (connectToInterfaces.Contains(typeof(TConnectFrom)))
+            {
+                return true;
+            }
+
```

Comment:
Lines 1439-1453 should just go away. Instead we just need to add a variation of lines 1439-1443 (on the LEFT) with the From and To reversed.

Reply:
Done.